File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-15.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) { 
        if (!root) return [] 

        const q = [root] 
        const result = [] 

        while(q.length) {
            const num = q.length 
            let level = []
            for (let i = 0; i < num; i ++) {
                const node = q.shift() 
                level.push(node.val) 
                if (node.left) q.push(node.left) 
                if (node.right) q.push(node.right)
            }  

            if (level.length) {
                result.push(level)
            }
        } 

        return result
    } 
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-15.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        def dfs(node, curMax): 
            if not node: 
                return 0 

            count = 1 if node.val >= curMax else 0 
            nextMax = max(node.val, curMax) 

            return count + dfs(node.left, nextMax) + dfs(node.right, nextMax) 

        return dfs(root, float('-inf')) 
